content/posts/csv-boilerplate-for-go.md

```diff
@@ -0,0 +1,307 @@
+
+Of course we still need to update our `UnmarshalRecord` and `MarshalRecord` from earlier. The good news is that we're unlikely to ever have to modify this code again:
+
+```go
+func (c *Car) UnmarshalRecord(record []string) error {
+	if got, want := len(record), len(carColumns); got != want {
+		return fmt.Errorf("bad number of columns: got=%d want=%d", got, want)
+	}
+	for i, col := range carColumns {
+		if err := col.UnmarshalValue(c, record[i]); err != nil {
+			return fmt.Errorf("column=%q: %w", col.Name, err)
+		}
+	}
+	return nil
+}
+
+func (c *Car) MarshalRecord() ([]string, error) {
+	record := make([]string, len(carColumns))
+	for i, col := range carColumns {
+		val, err := col.MarshalValue(c)
+		if err != nil {
+			return nil, err
+		}
+		record[i] = val
+	}
+	return record, nil
+}
+```
+
+Now it's time to implement the CSV decoding itself. My boilerplate implements this via the `ReadCarsCSV` function that can be seen below.
+
+```go
+const carComma = ';'
+
+func ReadCarsCSV(r io.Reader) ([]*Car, error) {
+	cr := csv.NewReader(r)
+	cr.Comma = carComma
+
+	records, err := cr.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	var cars []*Car
+	for i, record := range records {
+		if got, want := len(record), len(carColumns); got != want {
+			return nil, fmt.Errorf("row=%d: bad number of columns: got=%d want=%d", i+1, got, want)
+		}
+
+		switch i {
+		case 0:
+			for i, got := range record {
+				if want := carColumns[i].Name; got != want {
+					return nil, fmt.Errorf("unexpected header column %d: got=%q want=%q", i, got, want)
+				}
+			}
+		case 1:
+			for i, got := range record {
+				if want := carColumns[i].Type; got != want {
+					return nil, fmt.Errorf("unexpected type column %d: got=%q want=%q", i, got, want)
+				}
+			}
+		default:
+			car := &Car{}
+			if err := car.UnmarshalRecord(record); err != nil {
+				return nil, fmt.Errorf("row=%d: %w", i+1, err)
+			}
+			cars = append(cars, car)
+		}
+	}
+
+	return cars, nil
+}
+```
+
+The code above accomplishes the main task of returning all cars read from the given `io.Reader`, but it also validates the number of columns, as well as the header and type names found in the first two rows. Error messages should also be good, including both the row number as well the the offending column name in case something goes wrong for one of the records.
+
+It should also be easy to modify. For example if you want to ignore the second header line during reading, just remove the code. Or if instead of requiring all headers to have a fixed position, you could dynamically discover their position from the input by matching their names against the names in the `carColumns` slice and then reorder the elements in the record accordingly.
```

Comment:
```suggestion
It should also be easy to modify. For example if you want to ignore the second header line during reading, just remove the code. Or instead of requiring all headers to have a fixed position, you could dynamically discover their position from the input by matching their names against the names in the `carColumns` slice and then reorder the elements in the record accordingly.
```